asmcomp/cmm_helpers.ml

```diff
@@ -224,6 +223,38 @@ let asr_int c1 c2 dbg =
   | _ ->
       Cop(Casr, [c1; c2], dbg)
 
+let add_float c1 c2 dbg =
+  Cop(Caddf, [c1; c2], dbg)
+
+let sub_float c1 c2 dbg =
+  Cop(Csubf, [c1; c2], dbg)
+
+let mul_float c1 c2 dbg =
+  match (c1, c2) with
+  | (c, Cconst_float(2.0, _)) | (Cconst_float(2.0, _), c) ->
+     (* We transform x *. 2.0 into x +. x, by creating a temporary variable and
+        outputting let tmp = x in tmp +. tmp to avoid calculating x twice. *)
+     let tmp_var = V.create_local "tmp" in
+     let var = Cvar tmp_var in
+     Clet (VP.create tmp_var, c, add_float var var dbg)
+  | _, _ ->
+      Cop(Cmulf, [c1; c2], dbg)
+
+let div_float c1 c2 dbg =
+  match (c1, c2) with
+  | c, Cconst_float(f, _) ->
+      let x, exp = Float.frexp f in (* x = 0.5 if and only if f is a power of 2 *)
+      (* We can transform x/.2^{N} into x*.2^{-N} if and only if |N| < 1023,
+         because otherwise one of them may not be a valid floating point number
+         (2.**1023. evaluates to +infinity). Here, N=exp-1. *)
+      if x = 0.5 && abs exp - 1 < 1023
```

Comment:
I cannot make sense of this test. Did you mean `abs (exp - 1) < 1023` instead?

Moreover, `2**1023` is not infinite (but `2**1024` is).

Finally, while we wrote `|N| < 1023` in the JAR paper, I wonder if it is a typo. It seems to me that `|N| <= 1023` would work just as well.

Reply:
Sorry the `abs exp - 1` is a typo, because I originially wrote `abs exp`.

Reply:
I made a commit to take into account this modification.